### PR TITLE
EVM: Implement getBlockByNumber query

### LINKED ISSUE #878: EVM: Implement `eth_getBlockByNumber` endpoint.
## Summary
Implement the endpoint according to the Ethereum spec.

### PR DESCRIPTION
**Pull request by orkunkilic**
# Description
Implements get_block_by_number method in sov-evm query. Implemented "earliest", "latest", and hex parameters. Implemented both details(full / hash) on transaction return type.

Block_number defaults to latest block if not present.

Looking forward for efficiency feedbacks!

## Linked Issues
- Fixes #878
- Related to #502 

## Testing
Describe how these changes were tested. If you've added new features, have you added unit tests?
Tests WIP

## Docs
Describe where this code is documented. If it changes a documented interface, have the docs been updated?
No docs added yet.

### CODE REVIEW

File: module-system/module-implementations/sov-evm/src/query.rs
Comment (LukaszRozmej) [inline module-system/module-implementations/sov-evm/src/query.rs]: `head` might not be available in this context, nor it will have `sealed` block -> please use `blocks.last()`

File: module-system/module-implementations/sov-evm/src/query.rs
Comment (LukaszRozmej) [inline module-system/module-implementations/sov-evm/src/query.rs]: I think we won't support `pending` as it doesn't make sense for us, for `pending` block we would have to get the next DA block.

File: module-system/module-implementations/sov-evm/src/query.rs
Comment (LukaszRozmej) [inline module-system/module-implementations/sov-evm/src/query.rs:38]: this can fail and instead of panic, we need to return a proper error to rpc.

File: module-system/module-implementations/sov-evm/src/query.rs
Comment (LukaszRozmej) [inline module-system/module-implementations/sov-evm/src/query.rs]: `blocks.last` like above

File: module-system/module-implementations/sov-evm/src/query.rs
Comment (LukaszRozmej) [inline module-system/module-implementations/sov-evm/src/query.rs]: we shouldn't iterate over all transactions_hashes, that will load the entire database!

File: module-system/module-implementations/sov-evm/src/query.rs
Comment (LukaszRozmej) [inline module-system/module-implementations/sov-evm/src/query.rs]: same here we cannot load entire database.

File: module-system/module-implementations/sov-evm/src/query.rs
Comment (LukaszRozmej) [inline module-system/module-implementations/sov-evm/src/query.rs]: This code should be extracted to helper function it will be used in many RPC calls.

**Review by LukaszRozmej — DISMISSED** [review]

File: module-system/module-implementations/sov-evm/src/query.rs
Comment (orkunkilic) [reply module-system/module-implementations/sov-evm/src/query.rs]: I am not sure about that. There will be some rollups that use single centralized sequencer to give soft confirmations, such as ours. In that case, there will be a pending block on the sequencer level which is not broadcasted or sent to the DA layer.

File: module-system/module-implementations/sov-evm/src/query.rs
Comment (orkunkilic) [reply module-system/module-implementations/sov-evm/src/query.rs]: That was my mistake, I confused block.transactions with db transactions. Fixed.

File: module-system/module-implementations/sov-evm/src/query.rs
Comment (LukaszRozmej) [reply module-system/module-implementations/sov-evm/src/query.rs]: ok but then it should live on the sequencer and not on the module. Sequencer can extend the api, it will be a separate component that can us the module.. I don't see this doable inside the module.

**Review by LukaszRozmej — COMMENTED** [review]
needs a test

File: module-system/module-implementations/sov-evm/src/query.rs
Comment (bkolad) [reply module-system/module-implementations/sov-evm/src/query.rs]: I think we need to merge [#870](https://github.com/Sovereign-Labs/sovereign-sdk/pull/870) before writing a test. [@orkunkilic](https://github.com/orkunkilic) let's discuss the use case you have on our sync-call. I am sure we will accommodate it. In the meantime, I'm ok with two options: we can either merge this PR now and make a comment on #878 that we need to further work on the `test`  (once #870 lands) and the `pending block` case, or we can wait to discuss it during our upcoming sync-up call, whichever you prefer, [@orkunkilic](https://github.com/orkunkilic). Thank you for your excellent contribution!

Comment (LukaszRozmej) [reply module-system/module-implementations/sov-evm/src/query.rs]: I think you can write test without it, I've written receipts and transactions tests without it. This should be similar In full node it should work. 

But also depends if it should be unit test on evm module or integration test on full rollup.

**Comment by orkunkilic** [comment]
Tests are added, I think good to go after final review.

**Review by bkolad — APPROVED** [review]

File: module-system/module-implementations/sov-evm/src/query.rs
Comment (LukaszRozmej) [inline module-system/module-implementations/sov-evm/src/query.rs:89]: You shouldn't need to collect it here, this causes unnecessary allocation on heap.

**Pull request merged**
